### PR TITLE
Semaphore without probes: dummy notes

### PR DESCRIPTION
**Pull request by gretay-js**
In the original version of probes (before PR #60), inlining of `[%probe_is_enabled" <name>]` into a another compilation unit can result in link error if there is no mention of the `[%probe  <name>]` with the same `<name>` in the target-of-inlining compilation unit. For example:
```
$ cat t1.ml
let () = Printf.printf "%%probe_is_enabled = %b\n" (T2.probe_is_enabled ());

$ cat t2.ml
let [@inline always] probe_is_enabled () = [%probe_is_enabled "test2"]

$ ocamlopt.opt t2.ml t1.ml -o t.exe
t1.o: In function `camlT1__entry':
:(.text+0x7): undefined reference to `camlT1__semaphore_test2'
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
EXIT STATUS 2
```
This PR fixes it by emitting "dummy" probe sites.  (PR#60 fixes this problem as well by creating a unique semaphore label per name.)

Additionally, this PR removes the source-level requirement on compilation unit that has `[%probe_is_enabled <name>]` to contain also `[%probe <name>]`.  

There is a small downside to removing this check, which can catch some spelling mistakes in probe names that would otherwise go unnoticed. This is not catching all such issues, and anyway most `[%probe ]` commands are expected to be generated automatically by a ppx and so won't have these issues.

The main motivation for supporting `[%probe_is_enabled <name>]` without any `[%probe <name>]` in the source of the program is to toggle tracing/logging code that is not inside a `[%probe]` at runtime from another process with low overhead, compared to RPC.  A corresponding change in probes library will ensure that dummy probe sites are not updated. As a result, semaphore-only "probes" will be toggled without pausing the tracee. Systemtap will continue to work as before in the presence of dummy probes.

Dummy probes are implemented by adding special (unused) function with name ending in `___dummy_probes` per compilation unit if needed, with a `NOP` instruction for each dummy probe site and without a cmp/call instruction to the probe wrapper. 

It is a small change after refactoring of probe type (that is by itself an improvement).  Therefore, this RP is best reviewed commit by commit. 

It is necessary to advance the version of probe notes, because old tracer tool must not attempt to update dummy probe sites, which do not have an ocaml probe handler.

I'm not sure if this change needs bootstrap.

### CODE REVIEW

**Comment by gretay-js** [comment]
This design is based on discussion with @trishume.

**Comment by gretay-js** [comment]
I've rebased it and removed the bootstrap commit (it doesn't seem to be needed). This PR is now ready for review, but I haven't tested it with corresponding changes in probes_lib.

File: backend/amd64/emit.mlp [hunk 1]
```diff
@@ -1345,16 +1352,17 @@ let emit_probe_handler_wrapper p =
...
   (* Load probe arguments to correct locations for the handler *)
   Array.iteri (fun i reg -> move_allowing_stack_to_stack tmp.(i) reg) loc_args;
   (* Reload spilled registers used as temporaries *)
   I.mov (reg saved_r15) r15;
   (* Emit call to handler *)
-  add_used_symbol handler_code_sym;
-  emit_call handler_code_sym;
+  let h = Option.get p.probe_handler_code_sym in
+  add_used_symbol h;
+  emit_call h;
```
Comment (xclerc) [inline backend/amd64/emit.mlp:1365]: (This change is probably not needed.)

File: backend/amd64/emit.mlp [hunk 2]
```diff
@@ -1283,23 +1293,20 @@ let make_stack_loc ~offset i (r : Reg.t) =
...
   for i = 0 to Proc.num_register_classes - 1 do
-    num_stack_slots.(i) <- p.num_stack_slots.(i);
+    num_stack_slots.(i) <- p.probe_num_stack_slots.(i);
   done;
   (* Account for the return address that is now pushed on the stack. *)
   stack_offset := !stack_offset + 8;
   (* Emit function entry code *)
-  D.comment (Printf.sprintf "probe %s %s" probe_name handler_code_sym);
+  D.comment (Printf.sprintf "probe %s %s" p.probe_name
+               handler_code_sym);
```
Comment (xclerc) [inline backend/amd64/emit.mlp:1309]: (Does not look particularly long on one line.)

File: backend/amd64/emit.mlp [hunk 3]
```diff
@@ -1383,7 +1391,7 @@ let emit_probe_handler_wrapper p =
   cfi_endproc ();
   emit_function_type_and_size wrap_label
 
-let emit_probe_notes0 () =
+let emit_probe_notes0 probes =
   begin match system with
   | S_gnu | S_cygwin | S_solaris | S_win32 | S_linux_elf | S_bsd_elf
   | S_beos | S_mingw | S_win64 | S_linux | S_mingw64 | S_unknown ->
```
Comment (xclerc) [inline backend/amd64/emit.mlp:1397]: (Not part of this pull request, but this seems a bit
wide; shouldn't _e.g._ Windows and unknown have
their own case -- doing nothing? If changed, beware
of other occurences.)

File: backend/amd64/emit.mlp [hunk 4]
```diff
@@ -1475,10 +1478,56 @@ let emit_probe_notes0 () =
...
+        let probe = { probe_name;
+                      probe_label;
+                      probe_arg = [||];
+                      probe_live = Reg.Set.empty;
+                      probe_handler_code_sym = None;
+                      (* the following fields are unused for dummy probes,
+                         which have no ocaml handler and do not call
+                         probe wrappers. *)
+                      probe_stack_offset = !stack_offset;
+                      probe_num_stack_slots = num_stack_slots;
```
Comment (xclerc) [inline backend/amd64/emit.mlp:1519]: (If they are not used, maybe give them nonsensical values.
Another possibility would be to use different types, but that
seems overkill.)

File: backend/amd64/emit.mlp [hunk 5]
```diff
@@ -1440,13 +1443,13 @@ let emit_probe_notes0 () =
       D.qword (const 0)
     end;
     D.qword (ConstLabel semaphore_label);
-    D.bytes "ocaml.1\000";
-    D.bytes (probe_name ^ "\000");
+    D.bytes "ocaml.2\000";
```
Comment (xclerc) [inline backend/amd64/emit.mlp:1446]: (Is this something a user might want to query for some
reason? In other words, should this be moved to _e.g._
`Config` and be printed when the compiler is invoked
with `-config`?)

**Review by xclerc — APPROVED** [review]

File: backend/amd64/emit.mlp (hunk 5 above)
Comment (gretay-js) [reply backend/amd64/emit.mlp:1446]: Yes, it could be useful, thanks for the suggestion! Probes already have a variable in the Config, I can change it from boolean to probes_version option or something. I'll make a separate PR for it.

File: backend/amd64/emit.mlp (hunk 1 above)
Comment (gretay-js) [reply backend/amd64/emit.mlp]: fixed, thanks!

File: backend/amd64/emit.mlp (hunk 2 above)
Comment (gretay-js) [reply backend/amd64/emit.mlp]: fixed.

File: backend/amd64/emit.mlp (hunk 3 above)
Comment (gretay-js) [reply backend/amd64/emit.mlp:1395]: At configure time, probes are disabled on all windows-based systems, so the corresponding cases won't be reachable.

File: backend/amd64/emit.mlp (hunk 4 above)
Comment (gretay-js) [reply backend/amd64/emit.mlp:1517]: Yes, proper types seem to over-complicate the rest of the probes emitter code for a minor case.  I can't think of a "safe" invalid value for stack_offset, so I'll the code as is for now.

Comment (xclerc) [reply backend/amd64/emit.mlp:1517]: (My point was precisely an unsafe/notable value, so that
any mistake/misuse is obvious; a `deadbeef`-like value
if you will.)

**Comment by gretay-js** [comment]
It turns out that #60 did not fix all the instances of the link error mentioned in the description. 
So the change in this PR is needed even when %probe is in the same file as %probe_is_enable.

With flambda, the following example program fails to link:
```
$ cat t.ml
let _unused () = [%probe "foo" ()]
let is_enabled () = [%probe_is_enabled "foo"]

$ cat t.mli
val is_enabled : unit -> bool

$ cat s.ml
let () =   Printf.printf "%b\n" (T.is_enabled ())

$ocamlopt.opt t.mli t.ml s.ml -o -t.exe
s.o: In function `camlS__entry':
:(.text+0x7): undefined reference to `caml_probes_semaphore_foo'
t.o: In function `camlT__is_enabled_18':
:(.text+0x3): undefined reference to `caml_probes_semaphore_foo'
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
Error: Error during linking (exit code 1)
```
Thanks to @trishume  for the example.

**Review by lpw25 — APPROVED** [review]
Approving the parts that touch typing
